core/src/repair_weight.rs

```diff
@@ -245,29 +282,62 @@ impl RepairWeight {
         repairs
     }
 
-    /// Split `slot` and descendants into an orphan tree in repair weighting
-    /// These orphaned slots should be removed from `unrooted_slots` as on proper repair these slots might
-    /// now be part of the rooted path
+    /// Split `slot` and descendants into an orphan tree in repair weighting.
+    ///
+    /// Even if these orphaned slots were previously pruned, they should be added back to
+    /// `self.trees` as we are no longer sure of their ancestory. After they are repaired there is
+    /// a chance that they  are now part of the rooted path, when the ancestory information is
+    /// updated. If not they will once again be attached to the pruned set in
+    /// `update_orphan_ancestors`.
+    ///
+    /// Assumes that `slot` is greater than `self.root`.
     pub fn split_off(&mut self, slot: Slot) {
+        assert!(slot >= self.root);
         if slot == self.root {
             error!("Trying to orphan root of repair tree {}", slot);
             return;
         }
-        if let Some(subtree_root) = self.slot_to_tree.get(&slot) {
-            if *subtree_root == slot {
-                info!("{} is already orphan, skipping", slot);
-                return;
+        match self.slot_to_tree.get(&slot).copied() {
+            Some(TreeRoot::Root(subtree_root)) => {
+                if subtree_root == slot {
+                    info!("{} is already orphan, skipping", slot);
+                    return;
+                }
+                let subtree = self
+                    .trees
+                    .get_mut(&subtree_root)
+                    .expect("`self.slot_to_tree` and `self.trees` must be in sync");
+                let orphaned_tree = subtree.split_off(&(slot, Hash::default()));
+                self.rename_tree_root(&orphaned_tree, TreeRoot::Root(slot));
+                self.trees.insert(slot, orphaned_tree);
             }
-            let subtree = self
-                .trees
-                .get_mut(subtree_root)
-                .expect("subtree must exist");
-            let orphaned_tree = subtree.split_off(&(slot, Hash::default()));
-            for ((orphaned_slot, _), _) in orphaned_tree.all_slots_stake_voted_subtree() {
-                self.unrooted_slots.remove(orphaned_slot);
-                self.slot_to_tree.insert(*orphaned_slot, slot);
+            Some(TreeRoot::PrunedRoot(subtree_root)) => {
+                info!(
+                    "Dumping pruned slot {} of tree {} in repair",
+                    slot, subtree_root
+                );
+                let mut subtree = self
+                    .pruned_trees
+                    .remove(&subtree_root)
+                    .expect("`self.slot_to_tree` and `self.pruned_trees` must be in sync");
+
+                if subtree_root == slot {
+                    // In this case we simply unprune the entire subtree
```

Comment:
Maybe a definition of `unprune` would be helpful here, `unprune the entire subtree by adding this subtree back into the main set of trees, self.trees`